src/ariel-os-coap/src/stored.rs

```diff
@@ -0,0 +1,124 @@
+//! Credential and key configuration backed by ariel-os storage
+
+use ariel_os_debug::log::{debug, info};
+use cbor_macro::cbo;
+use coapcore::seccfg::ServerSecurityConfig;
+
+mod flash_peers {
+    include!(concat!(env!("OUT_DIR"), "/peers.rs"));
+}
+
+pub async fn server_security_config() -> impl ServerSecurityConfig {
+    StoredPolicy::load().await
+}
+
+// While this might be a ZST eventually, right now we're loading all this at startup because we
+// don't have the async context to access any storage at CoAP time.
+struct StoredPolicy {
+    own_edhoc_credential: (lakers::Credential, lakers::BytesP256ElemLen),
+}
+
+impl ServerSecurityConfig for StoredPolicy {
+    // FIXME: Decide based on peers.rs input (but so far tokens are not implemented)
+    const PARSES_TOKENS: bool = false;
+    const HAS_EDHOC: bool = true;
+    type GeneralClaims = StoredClaims;
+
+    fn own_edhoc_credential(&self) -> Option<(lakers::Credential, lakers::BytesP256ElemLen)> {
+        Some(self.own_edhoc_credential)
+    }
+
+    fn expand_id_cred_x(
+        &self,
+        id_cred_x: lakers::IdCred,
+    ) -> Option<(lakers::Credential, StoredClaims)> {
+        for (credential, scope) in flash_peers::kccs() {
+            if credential.by_kid().is_ok_and(|by_kid| by_kid == id_cred_x)
+                || credential
+                    .by_value()
+                    .is_ok_and(|by_value| by_value == id_cred_x)
+            {
+                return Some((credential, StoredClaims { scope }));
+            }
+        }
+        None
+    }
+}
+
+impl StoredPolicy {
+    async fn load() -> Self {
+        // Storage format: ([u8], [u8; 32]), where the former is a CCS, and the latter the
+        // corresponding key. We may need to extend the latter to be a COSE_Key when crypto agility
+        // becomes a thing.
+        const OWN_CREDENTIAL_KEY: &str = "coapcore.own-edhoc-credential";
+
+        let mut storage = ariel_os_storage::lock().await;
+        let (credential, key) = match storage
+            .get(OWN_CREDENTIAL_KEY)
+            .await
+            .expect("flash error prevents startup")
+        {
+            Some(credpair) => credpair,
+            None => {
+                use lakers::CryptoTrait;
```

Comment:
Style wise I find it off to have the generation, and serialization, of the keypair within a function that is called `load()`, while holding a lock.
Maybe split out some helper functions?

Reply:
Good point, will do. (Held the lock because it is a read-or-set-default thing, but realistically there won't be parallel calls, and there is no harm as long as there is one that is eventually set).